pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/MultiStageOperator.java

```diff
@@ -32,14 +32,14 @@
 public abstract class MultiStageOperator implements Operator<TransferableBlock>, AutoCloseable {
   private static final org.slf4j.Logger LOGGER = LoggerFactory.getLogger(MultiStageOperator.class);
 
-  private final String _operatorId;
-  private final OpChainExecutionContext _context;
+  private final OperatorId _operatorId;
+  protected final OpChainExecutionContext _context;
   protected final OpChainStats _opChainStats;
 
   public MultiStageOperator(OpChainExecutionContext context) {
     _context = context;
-    _operatorId =
-        Joiner.on("_").join(toExplainString(), _context.getRequestId(), _context.getStageId(), _context.getServer());
+    _operatorId = new OperatorId(_context.getRequestId(), _context.getStageId(), _context.getServer().toString(),
+        toExplainString(), null);
```

Comment:
it seems like the operatorIndex is always Null here? can't we parsed this into OpChainExecutionContext?

Reply:
Yeah, so issue is the opChainContext is created during the visitor. And before visiting all nodes, we can't really index them correctly based on the position in the chain.